File: changji/src/changji/models/project.py
"""项目模型与磁盘布局。

一个项目就是一个自包含的目录，换机器整个拷走即可。目录里只有相对路径，
不含任何绝对路径，也不含模型文件（那些跟机器走，不跟项目走）。

    我的短剧/
    ├── project.json          项目元数据与分镜表
    ├── assets.json           角色与场景资产库
    ├── changji.toml          项目级配置覆盖（可选）
    ├── refs/                 角色三视图、场景空景图
    ├── audio/                配音
    ├── frames/               逐镜首帧
    ├── shots/
    │   ├── draft/            草稿档视频
    │   └── final/            成片档视频
    ├── subtitles/
    └── output/               成片
"""

from __future__ import annotations

import json
import shutil
import tempfile
from datetime import datetime, timezone
from pathlib import Path

from pydantic import BaseModel, Field

from .character import AssetLibrary, StyleLine
from .shot import Shot, ShotStatus

PROJECT_FILE = "project.json"
ASSETS_FILE = "assets.json"
SCHEMA_VERSION = 1

SUBDIRS = (
    "refs", "audio", "frames", "shots/draft", "shots/final",
    "subtitles", "output", "logs",
)


class Episode(BaseModel):
    """一集。分镜表挂在这里。"""

    model_config = {"extra": "forbid"}

    episode_id: str = Field(pattern=r"^[a-z0-9_]+$")
    title: str = ""
    synopsis: str = ""
    target_duration_s: float = Field(default=180.0, gt=0, description="目标时长")
    script: str = Field(default="", description="剧本原文")
    shots: list[Shot] = Field(default_factory=list)

    def sorted_shots(self) -> list[Shot]:
        return sorted(self.shots, key=lambda s: s.order)

    def shot_by_id(self, shot_id: str) -> Shot | None:
        return next((s for s in self.shots if s.shot_id == shot_id), None)

    def planned_duration_s(self) -> float:
        return sum(s.duration_s for s in self.shots)

    def counts_by_status(self) -> dict[str, int]:
        out: dict[str, int] = {}
        for s in self.shots:
            out[s.status.value] = out.get(s.status.value, 0) + 1
        return out

    def shots_needing(self, status: ShotStatus) -> list[Shot]:
        """取处于某个阶段的镜头。断点续跑靠它。"""
        return [s for s in self.sorted_shots() if s.status == status]


class Project(BaseModel):
    """一个项目。"""

    model_config = {"extra": "forbid"}

    schema_version: int = SCHEMA_VERSION
    project_id: str = Field(pattern=r"^[a-z0-9_-]+$")
    title: str = ""
    style_line: StyleLine = StyleLine.REALISTIC
    # 这部剧讲什么。写下一集时当提示词用。
    # 不存的话，隔天想接着写第六集，得凭记忆把当初那句话重打一遍。
    premise: str = Field(default="", max_length=2000)
    created_at: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    episodes: list[Episode] = Field(default_factory=list)

    def episode_by_id(self, episode_id: str) -> Episode | None:
        return next((e for e in self.episodes if e.episode_id == episode_id), None)

    def touch(self) -> None:
        self.updated_at = datetime.now(timezone.utc).isoformat()


class ProjectPaths:
    """项目目录布局。所有路径都由项目根推导，绝不写死。"""

    def __init__(self, root: str | Path) -> None:
        self.root = Path(root).expanduser().resolve()

    def ensure(self) -> ProjectPaths:
        for sub in SUBDIRS:
            (self.root / sub).mkdir(parents=True, exist_ok=True)
        return self

    @property
    def project_file(self) -> Path:
        return self.root / PROJECT_FILE

    @property
    def assets_file(self) -> Path:
        return self.root / ASSETS_FILE

    @property
    def refs(self) -> Path:
        return self.root / "refs"

    @property
    def audio(self) -> Path:
        return self.root / "audio"

    @property
    def frames(self) -> Path:
        return self.root / "frames"

    @property
    def subtitles(self) -> Path:
        return self.root / "subtitles"

    @property
    def output(self) -> Path:
        return self.root / "output"

    @property
    def logs(self) -> Path:
        return self.root / "logs"

    def shots(self, tier: str) -> Path:
        return self.root / "shots" / tier

    def rel(self, path: str | Path) -> str:
        """绝对路径转成相对项目根的路径。存进 JSON 的一律用这个。

        用正斜杠，保证在 Windows 上存的项目拿到 Linux 上也能读。
        """
        p = Path(path).expanduser().resolve()
        try:
            return p.relative_to(self.root).as_posix()
        except ValueError:
            # 不在项目内的文件只能存绝对路径，但这会破坏可移植性
            raise ValueError(
                f"路径不在项目目录内，存进项目会破坏可移植性：{p}\n"
                f"请先把文件复制进 {self.root}"
            ) from None

    def abs(self, rel_path: str) -> Path:
        """相对路径还原成绝对路径。"""
        return (self.root / rel_path).resolve()


class ProjectStore:
    """项目的读写。写入用原子替换，避免中途断电留下半个文件。"""

    def __init__(self, root: str | Path) -> None:
        self.paths = ProjectPaths(root)

    @property
    def root(self) -> Path:
        return self.paths.root

    def exists(self) -> bool:
        return self.paths.project_file.is_file()

    # ---- 创建 ----

    @classmethod
    def create(
        cls,
        root: str | Path,
        project_id: str,
        title: str = "",
        style_line: StyleLine = StyleLine.REALISTIC,
    ) -> ProjectStore:
        store = cls(root)
        if store.exists():
            raise FileExistsError(f"这个目录已经是一个项目了：{store.root}")
        store.paths.ensure()
        project = Project(project_id=project_id, title=title or project_id,
                          style_line=style_line)
        store.save_project(project)
        store.save_assets(AssetLibrary(style={"style_line": style_line}))
        return store

    # ---- 读 ----

    def load_project(self) -> Project:
        if not self.exists():
            raise FileNotFoundError(
                f"这里不是一个项目目录：{self.root}\n"
                f"用 changji new 创建，或者 cd 到正确的目录"
            )
        raw = self._read_json(self.paths.project_file)
        version = raw.get("schema_version", 0)
        if version > SCHEMA_VERSION:
            raise ValueError(
                f"项目是用更新版本的场记创建的（格式版本 {version}，"
                f"本机支持到 {SCHEMA_VERSION}）。请升级后再打开"
            )
        return Project.model_validate(raw)

    def load_assets(self) -> AssetLibrary:
        if not self.paths.assets_file.is_file():
            return AssetLibrary()
        return AssetLibrary.model_validate(self._read_json(self.paths.assets_file))

    # ---- 写 ----

    def save_project(self, project: Project) -> None:
        project.touch()
        self._write_json(self.paths.project_file, project.model_dump(mode="json"))

    def save_assets(self, assets: AssetLibrary) -> None:
        self._write_json(self.paths.assets_file, assets.model_dump(mode="json"))

    # ---- 内部 ----

    @staticmethod
    def _read_json(path: Path) -> dict:
        try:
            return json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            raise ValueError(f"文件损坏，不是合法 JSON：{path}\n{exc}") from exc

    @staticmethod
    def _write_json(path: Path, data: dict) -> None:
        """原子写。先写临时文件再替换，中途断电不会留下半个文件。"""
        path.parent.mkdir(parents=True, exist_ok=True)
        fd, tmp = tempfile.mkstemp(dir=str(path.parent), suffix=".tmp")
        try:
            with open(fd, "w", encoding="utf-8") as fh:
                json.dump(data, fh, ensure_ascii=False, indent=2)
                fh.flush()
            Path(tmp).replace(path)
        except BaseException:
            Path(tmp).unlink(missing_ok=True)
            raise

    # ---- 迁移 ----

    def copy_into(self, src: str | Path, subdir: str, name: str | None = None) -> str:
        """把外部文件复制进项目，返回相对路径。

        参考图这类素材必须复制进来而不是引用原位置，否则项目拷到别的机器就断链。
        """
        source = Path(src).expanduser().resolve()
        if not source.is_file():
            raise FileNotFoundError(f"文件不存在：{source}")
        target_dir = self.root / subdir
        target_dir.mkdir(parents=True, exist_ok=True)
        target = target_dir / (name or source.name)
        if source != target:
            shutil.copy2(source, target)
        return self.paths.rel(target)

File: changji/webapp/server/src/engine/models/project.js
/**
 * 项目模型与磁盘布局。
 *
 * 一个项目就是一个自包含的目录，换机器整个拷走即可。目录里只有相对路径，
 * 不含任何绝对路径，也不含模型文件（那些跟机器走，不跟项目走）。
 *
 *     我的短剧/
 *     ├── project.json          项目元数据与分镜表
 *     ├── assets.json           角色与场景资产库
 *     ├── changji.toml          项目级配置覆盖（可选）
 *     ├── refs/                 角色三视图、场景空景图
 *     ├── audio/                配音
 *     ├── frames/               逐镜首帧
 *     ├── shots/
 *     │   ├── draft/            草稿档视频
 *     │   └── final/            成片档视频
 *     ├── subtitles/
 *     └── output/               成片
 */

import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { z } from 'zod'

import { AssetLibrarySchema, StyleLine } from './character.js'
import { ShotSchema, ShotStatus } from './shot.js'

export const PROJECT_FILE = 'project.json'
export const ASSETS_FILE = 'assets.json'
export const SCHEMA_VERSION = 1

const SUBDIRS = [
  'refs', 'audio', 'frames', 'shots/draft', 'shots/final',
  'subtitles', 'output', 'logs',
]

/** 一集。分镜表挂在这里。 */
export const EpisodeSchema = z
  .object({
    episode_id: z.string().regex(/^[a-z0-9_]+$/),
    title: z.string().default(''),
    synopsis: z.string().default(''),
    target_duration_s: z.number().gt(0).default(180), // 目标时长
    script: z.string().default(''), // 剧本原文
    shots: z.array(ShotSchema).default([]),
  })
  .strict()

export const sortedShots = (episode) =>
  [...episode.shots].sort((a, b) => a.order - b.order)

export const shotById = (episode, shotId) =>
  episode.shots.find((s) => s.shot_id === shotId) ?? null

export const plannedDurationS = (episode) =>
  episode.shots.reduce((a, s) => a + s.duration_s, 0)

export function countsByStatus(episode) {
  const out = {}
  for (const s of episode.shots) out[s.status] = (out[s.status] ?? 0) + 1
  return out
}

/** 取处于某个阶段的镜头。断点续跑靠它。 */
export const shotsNeeding = (episode, status) =>
  sortedShots(episode).filter((s) => s.status === status)

export const ProjectSchema = z
  .object({
    schema_version: z.number().int().default(SCHEMA_VERSION),
    project_id: z.string().regex(/^[a-z0-9_-]+$/),
    title: z.string().default(''),
    style_line: z.enum(Object.values(StyleLine)).default(StyleLine.REALISTIC),
    // 这部剧讲什么。写下一集时当提示词用。
    // 不存的话，隔天想接着写第六集，得凭记忆把当初那句话重打一遍。
    premise: z.string().max(2000).default(''),
    created_at: z.string().default(() => new Date().toISOString()),
    updated_at: z.string().default(() => new Date().toISOString()),
    episodes: z.array(EpisodeSchema).default([]),
  })
  .strict()

export const episodeById = (project, episodeId) =>
  project.episodes.find((e) => e.episode_id === episodeId) ?? null

/** 项目目录布局。所有路径都由项目根推导，绝不写死。 */
export class ProjectPaths {
  constructor(root) {
    this.root = path.resolve(expandHome(String(root)))
  }

  ensure() {
    for (const sub of SUBDIRS) {
      fs.mkdirSync(path.join(this.root, sub), { recursive: true })
    }
    return this
  }

  get projectFile() {
    return path.join(this.root, PROJECT_FILE)
  }

  get assetsFile() {
    return path.join(this.root, ASSETS_FILE)
  }

  get refs() {
    return path.join(this.root, 'refs')
  }

  get audio() {
    return path.join(this.root, 'audio')
  }

  get frames() {
    return path.join(this.root, 'frames')
  }

  get subtitles() {
    return path.join(this.root, 'subtitles')
  }

  get output() {
    return path.join(this.root, 'output')
  }

  get logs() {
    return path.join(this.root, 'logs')
  }

  shots(tier) {
    return path.join(this.root, 'shots', tier)
  }

  /**
   * 绝对路径转成相对项目根的路径。存进 JSON 的一律用这个。
   *
   * 用正斜杠，保证在 Windows 上存的项目拿到 Linux 上也能读。
   */
  rel(target) {
    const p = path.resolve(expandHome(String(target)))
    const relative = path.relative(this.root, p)
    if (relative.startsWith('..') || path.isAbsolute(relative)) {
      // 不在项目内的文件只能存绝对路径，但这会破坏可移植性
      throw new Error(
        `路径不在项目目录内，存进项目会破坏可移植性：${p}\n请先把文件复制进 ${this.root}`,
      )
    }
    return relative.split(path.sep).join('/')
  }

  /** 相对路径还原成绝对路径。 */
  abs(relPath) {
    return path.resolve(this.root, relPath)
  }
}

function expandHome(p) {
  return p.startsWith('~') ? path.join(os.homedir(), p.slice(1)) : p
}

/** 项目的读写。写入用原子替换，避免中途断电留下半个文件。 */
export class ProjectStore {
  constructor(root) {
    this.paths = new ProjectPaths(root)
  }

  get root() {
    return this.paths.root
  }

  exists() {
    return fs.existsSync(this.paths.projectFile) &&
      fs.statSync(this.paths.projectFile).isFile()
  }

  // ---- 创建 ----

  static create(root, projectId, title = '', styleLine = StyleLine.REALISTIC) {
    const store = new ProjectStore(root)
    if (store.exists()) {
      const err = new Error(`这个目录已经是一个项目了：${store.root}`)
      err.code = 'EEXIST'
      throw err
    }
    store.paths.ensure()
    store.saveProject(
      ProjectSchema.parse({
        project_id: projectId,
        title: title || projectId,
        style_line: styleLine,
      }),
    )
    store.saveAssets(
      AssetLibrarySchema.parse({ style: { style_line: styleLine } }),
    )
    return store
  }

  // ---- 读 ----

  loadProject() {
    if (!this.exists()) {
      throw new Error(
        `这里不是一个项目目录：${this.root}\n用 changji new 创建，或者切到正确的目录`,
      )
    }
    const raw = readJson(this.paths.projectFile)
    const version = raw.schema_version ?? 0
    if (version > SCHEMA_VERSION) {
      throw new Error(
        `项目是用更新版本的场记创建的（格式版本 ${version}，本机支持到 ${SCHEMA_VERSION}）。请升级后再打开`,
      )
    }
    return ProjectSchema.parse(raw)
  }

  loadAssets() {
    if (!fs.existsSync(this.paths.assetsFile)) {
      return AssetLibrarySchema.parse({})
    }
    return AssetLibrarySchema.parse(readJson(this.paths.assetsFile))
  }

  // ---- 写 ----

  saveProject(project) {
    project.updated_at = new Date().toISOString()
    writeJson(this.paths.projectFile, project)
  }

  saveAssets(assets) {
    writeJson(this.paths.assetsFile, assets)
  }

  /**
   * 把外部文件复制进项目，返回相对路径。
   *
   * 参考图这类素材必须复制进来而不是引用原位置，否则项目拷到别的机器就断链。
   */
  copyInto(src, subdir, name = null) {
    const source = path.resolve(expandHome(String(src)))
    if (!fs.existsSync(source) || !fs.statSync(source).isFile()) {
      throw new Error(`文件不存在：${source}`)
    }
    const targetDir = path.join(this.root, subdir)
    fs.mkdirSync(targetDir, { recursive: true })
    const target = path.join(targetDir, name || path.basename(source))
    if (source !== target) fs.copyFileSync(source, target)
    return this.paths.rel(target)
  }
}

function readJson(file) {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'))
  } catch (err) {
    throw new Error(`文件损坏，不是合法 JSON：${file}\n${err.message}`)
  }
}

/** 原子写。先写临时文件再替换，中途断电不会留下半个文件。 */
function writeJson(file, data) {
  fs.mkdirSync(path.dirname(file), { recursive: true })
  const tmp = path.join(
    path.dirname(file),
    `.${path.basename(file)}.${process.pid}.${Date.now()}.tmp`,
  )
  try {
    fs.writeFileSync(tmp, JSON.stringify(data, null, 2), 'utf8')
    fs.renameSync(tmp, file)
  } catch (err) {
    try {
      fs.unlinkSync(tmp)
    } catch {
      // 临时文件本来就没建成，没什么可清的
    }
    throw err
  }
}

export { ShotStatus }
